orc/src/main/java/com/netflix/iceberg/orc/TypeConversion.java

```diff
@@ -0,0 +1,172 @@
+        }
+        return struct;
+      }
+      case LIST: {
+        Types.ListType list = (Types.ListType) type;
+        columnIds.add(list.elementId());
+        return TypeDescription.createList(toOrc(type.asListType().elementType(),
+            columnIds));
+      }
+      case MAP: {
+        Types.MapType map = (Types.MapType) type;
+        columnIds.add(map.keyId());
+        TypeDescription key = toOrc(map.keyType(), columnIds);
+        columnIds.add(map.valueId());
+        return TypeDescription.createMap(key,
+            toOrc(type.asMapType().valueType(), columnIds));
+      }
+      default:
+        throw new IllegalArgumentException("Unhandled type " + type.typeId());
+    }
+  }
+
+  /**
+   * Convert an ORC schema to an Iceberg schema.
+   * @param schema the ORC schema
+   * @param columnIds the column ids
+   * @return the Iceberg schema
+   */
+  public Schema fromOrc(TypeDescription schema, int[] columnIds) {
+    return new Schema(convertOrcToType(schema, columnIds).asStructType().fields());
+  }
+
+  Type convertOrcToType(TypeDescription schema, int[] columnIds) {
+    switch (schema.getCategory()) {
+      case BOOLEAN:
+        return Types.BooleanType.get();
+      case BYTE:
+      case SHORT:
+      case INT:
+        return Types.IntegerType.get();
+      case LONG:
+        return Types.LongType.get();
+      case FLOAT:
+        return Types.FloatType.get();
+      case DOUBLE:
+        return Types.DoubleType.get();
+      case STRING:
+      case CHAR:
+      case VARCHAR:
+        return Types.StringType.get();
+      case BINARY:
+        return Types.BinaryType.get();
+      case DATE:
+        return Types.DateType.get();
+      case TIMESTAMP:
+        return Types.TimestampType.withoutZone();
+      case DECIMAL:
+        return Types.DecimalType.of(schema.getPrecision(), schema.getScale());
+      case STRUCT: {
+        List<String> fieldNames = schema.getFieldNames();
+        List<TypeDescription> fieldTypes = schema.getChildren();
+        List<Types.NestedField> fields = new ArrayList<>(fieldNames.size());
+        for(int c=0; c < fieldNames.size(); ++c) {
+          String name = fieldNames.get(c);
+          TypeDescription type = fieldTypes.get(c);
+          fields.add(Types.NestedField.optional(columnIds[type.getId()], name,
+              convertOrcToType(type, columnIds)));
+        }
+        return Types.StructType.of(fields);
+      }
+      case LIST: {
+        TypeDescription child = schema.getChildren().get(0);
+        return Types.ListType.ofOptional(columnIds[child.getId()],
+            convertOrcToType(child, columnIds));
+      }
+      case MAP: {
+        TypeDescription key = schema.getChildren().get(0);
+        TypeDescription value = schema.getChildren().get(1);
+        switch (key.getCategory()) {
+          case STRING:
+          case CHAR:
+          case VARCHAR:
+            return Types.MapType.ofOptional(columnIds[key.getId()],
```

Comment:
Are keys required in ORC, or can they be null as well?

Reply:
Keys can be null.